PIMD: Fix pim_mroute_del crash while killing pimd.

patrasar opened the pull request:
When we kill pimd, pim_upstream_del() will be get called as part of cleaning process.
Delete mroute by calling pim_mroute_del() which in result initialize the up->channel_oil
to NULL after cleanup. Delete each ifchannel stored in the upstream ifchannel list one
by one. As part of pim_ifchannel_delete(), it will check if it is the last ifchannel,
then delete the corresponding upstream. So pim_upstream_del will be get called more than
once as part of pim_upstream_terminate() and pim_ ifchannel_delete(). Similarly pim_mroute_del()
will be again called and since there is no NULL check before accesing the data, crash is happening.

Fix:
Adding Null check in pim_mroute_del before accessing the data.
and using the FLAG "PIM_UPSTREAM_FLAG_DEL" which will check if the deletion of pim upstream is already
in progress then don’t call pim_upstream_del() from pim_ifchannel_delete().

Signed-off-by: Sarita Patra <saritap@vmware.com>

donaldsharp committed:
zebra: Allow a static daemon to connect

As part of moving the static route handling to it's own daemon
allow zebra to accept static route types from upper level
protocols.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
debianpkg, redhat, tools: Fixup startup/build for new daemon

Add code to allow FRR to properly build and handle the staticd
for some of the more common packaging.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
staticd: Start the addition of a staticd

This is the start of separating out the static
handling code from zebra -> staticd.  This will
help simplify the zebra code and isolate static
route handling to it's own code base.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
staticd: Enable backup config

When the user is not using the integrated config and
has upgraded to a version of FRR with staticd, that
probably means that static routes are stored in zebra.conf.
Take advantage of the backup backup config option and read
those in.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
zebra: Remove zebra_static.c and .h

These are no longer needed so remove.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
staticd: Fix recursive routes for certain types of nexthops

Existing NEXTHOP_TYPE_IPV4_IFINDEX and NEXTHOP_TYPE_IPV6_IFINDEX
routes allow recursion, but were broken when the route happened
to recursively resolve and the resolution nexthop changed.

This commit fixes this issue.  Please note that this issue was
in pre-move of static route handling to it's own daemon as well.
This was some easy low-hanging fruit, so to speak.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

rzalamena committed:
lib: import new JSON iteration macro

Save a few lines when iterating over JSON objects using the new
JSON_FOREACH macro.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rwestphal committed:
lib: add listnode_add_head()

Provide a new convenience function that adds an element to the beginning
of a list.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

donaldsharp committed:
*: ALLOC calls cannot fail

There is no need to check for failure of a ALLOC call
as that any failure to do so will result in a assert
happening.  So we can safely remove all of this code.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
lib: Remove memory check test

Found a missed memory check test that can just be removed.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
babeld: Use enum instead of #defines for kernel route type information

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
tests: Cleanup assumption that ALLOC could fail.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
doc: Add some documentation for staticd

Remove the ip route specific sections from zebra documenation and
create a specific one for the new staticd.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
staticd: Code review comments fixes

1) Conform staticd to proper gnu gpl file format.
2) Fix a spelling mistake found.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
lib, zebra: Allow protocols to use Distance as part of RR semantics

Allow protocols to specify to zebra that they would like zebra
to use the distance passed down as part of determine sameness for
Route Replace semantics.

This will be used by the static daemon to allow it to have
backup static routes with greater distances.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
staticd: Allow daemon to have backup static routes

Modify staticd to allow it to have backup static routes
with higher admin distance.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

rzalamena committed:
bfdd: imported new daemon source code

Import source code from external `bfdd` daemon ported from Cumulus PTM.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: wire to the build process

Add BFD daemon to the build process and packaging instructions.

Currently the bfdd daemon does nothing, this is just to document how the
daemon insertion step occured.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: add vty shell commands

Implement vty shell integration and allow `bfdd` to be configured
through FRR's vtysh.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: re-route PTM-BFD daemon messages

When `bfdd` is enabled - which it is by default - re-route the PTM-BFD
messages to the FRR's internal BFD daemon instead of the external
PTM daemon.

This will help the migration of BFD implementations and avoid
duplicating code.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bgpd/ospf(6)d/pimd: don't show BFD timers

When BFD timers are configured, don't show it anymore in the daemon
side. This will help us migrate the timers command from daemons to
`bfdd`.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: add documentation

Add BFD daemon documentation:
* commands;
* man page;
* manual / description;

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

donaldsharp committed:
staticd: Fix crash with NULL si

Move the call to the static_install_route to inside
of the loop, since we have changed the behavior of
how we send down routes to zebra a bit.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

rwestphal committed:
babeld: avoid reading the configuration file twice

The frr_config_fork() function called by all FRR daemons during
initialization already takes care of reading the startup configuration
file.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

rwestphal committed:
lib: implement DEFPY_NOSH

This is the DEFPY equivalent of DEFUN_NOSH.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

rwestphal committed:
zebra: fix "no pseudowire IFNAME" on vtysh

We must hide only "pseudowire IFNAME" from vtysh, the "no" form of the
command should be made available to the extract.pl script. Split the
command into two to fix this problem.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

rwestphal committed:
ldpd: use DEFPY_NOSH whenever possible

All ldpd commands were written using DEFPY except the ones that needed
to be ignored by vtysh, in which case we were using DEFUN_NOSH. Now that
DEFPY_NOSH is available, convert the remaining commands for consistency
and to simplify the code a little bit.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

rwestphal committed:
lib, vtysh: fix inconsistent VRF commands in vtysh

* Only zebra and pimd call vrf_cmd_init(), so these are the only daemons
  that should receive VRF commands from vtysh;
* "netns NAME" and "no netns NAME" are available only in zebra, write
  custom DEFSHs in vtysh to make it aware of that;
* Remove the "no vrf NAME" definition from vtysh.c and expose the
  original command to vtysh by converting the DEFUN_NOSH to a simple
  DEFUN. This command doesn't change the vty node so there's no need to
  special case it.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

rwestphal committed:
lib, vtysh: fix inconsistent interface commands in vtysh

The definition of the interface commands in vtysh.c were outdated.
Currently, all daemons that call if_cmd_init() will have the "no interface
IFNAME" command and the "[no] description" commands as well, so there's
no need to define exceptions for these commands anymore.

To fix this, make extract.pl parse the if.c file so that vtysh can get the
interface commands from there automatically. Only the "interface IFNAME
[vrf NAME]" must be kept in vtysh.c because it changes the vty node and
thus needs special treatment.

Finally, make pimd and pbrd display interface descriptions on "sh run"
when they are configured.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

rzalamena committed:
bfdd: show the correct transmission speed

When configured transmission speed doesn't match the actual speed, show
the difference in the output.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: smooth configuration transition

After configuring a new value set the polling bit to negotiate speeds
again next transmission cycle.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: show single hop local-address

Show local-address on single hop when configured.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: send replay request on zebra connection

This will make `bfdd` synchronize with its client when zebra dies or
bfdd is restarted.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: free zebra clients data on unregistration

Avoid a memory leak on client daemons restart by getting rid of old
registrations.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

qlyoung committed:
ripd: re-add check for rip_create fail

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

donaldsharp committed:
lib: Set proper write file when using backup config

When we read in a backup file, we should save the original
host.config so that we can put it back to the correct original
location after we read in the backup config.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

cfra committed:
isisd: add debug message if adjacency is ignored because IP is unusable

isisd verifies whether the neighboring IPv4 addresses overlap with its own
unless the interface is running in unnumbered mode. If no overlap is found
and IPv6 is also not enabled, IIHs will be ignored.

Add a debug message for this case, to avoid people wondering why adjacencies
are not coming up.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

cfra committed:
isisd: learn and advertise IPv6 dst-src routes

Receive IPv6 dst-src routes from zebra and advertise them in our LSPs
if so configured.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

cfra committed:
isisd: move route_table into spftree

As isisd's route_tables are directly related to spf trees, move
the route tables into the spftree instead of maintaining them
alongside of the spftrees.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

rzalamena committed:
bfdd: code refactory for simplification

Simplify code and remove unnecessary log messages. The old log messages
are going to be shown by the caller anyway.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: clean up header inclusion

Most of the headers we need are included by zebra.h, so lets simplify
this.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: improve logging messages

Show a little more details, remove some duplicated calls and remove the
macro compatibility with old debugging functions.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

rzalamena committed:
bfdd: clean-up bfd clients data on shutdown

On `zebra` / `bfdd` shutdown we now clean up all client data to avoid
memory leaks (ghost clients). This also prevents 'slow' shutdown on
`zebra` sparing us from seeing some rare topotests shutdown failures
(signal handler getting stopped by signal).

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

sworleys committed:
zebra: Ignore AF_ETHERNET and AF_MPLS in route_change as well

Remove AF_ETHERNET and AF_MPLS as valid address families for route
changes.

Signed-off-by: Stephen Worley <sworley@cumulusnetworks.com>

sworleys committed:
zebra: Format log for address family check

Switched the address family check to use unsigned format.

Signed-off-by: Stephen Worley <sworley@cumulusnetworks.com>

rzalamena committed:
bfdd: fix IPv6 peers using link-local address

When using link-local address we must specify the scope-id for the
address in order to bind to the interface.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

cfra committed:
isisd: fix redist_delete to also consider level-2

There was an off-by-one error in redist_delete, so that routes redistributed
into level-2 could never be withdrawn.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

cfra committed:
isisd: fix isis_route_merge_verify logic

This addresses two issues for L1L2 operation:

a) If an L1 route has ROUTE_ACTIVE unset and an L2 route for the same
destination has ROUTE_ACTIVE set, isisd would still put the L1 route
into the merged table. This causes the route for the destination to
get uninstalled from zebra until the next SPF run, which is incorrect.

To fix this, look at the ROUTE_ACTIVE flag and allow L2 routes to win
against L1 routes, when the L1 has ROUTE_ACTIVE unset.

b) If an L1 route wins against an existing L2 route, the ZEBRA_SYNCED
flag would remain on the L2 route. This leads to the problem that when
the L1 route disappears again, the L2 doesn't get reinstalled, since
isisd assumes it's already in the RIB because ZEBRA_SYNCED is set.

Solve this by clearing ZEBRA_SYNCED on L2 routes, if they lose against
an L1 route.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

cfra committed:
isisd: make use of advanced concepts like arrays and loops

Have an array of spftrees instead of a separate spftree and an
spftree6 for which all the code gets duplicated.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

cfra committed:
isisd: don't infer spftree from address family

Instead of using the address family to determine which spftree structure
should be used, specify it explicitly. With the advent of ipv6 dst-src
routing, the tree cannot be uniquely determined from the family.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

cfra committed:
isisd: make spf code dst-src aware

Take the source-prefix sub-TLV into consideration when running SPF
and support creation/deletion of dst-src routes as result.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

donaldsharp committed:
staticd, zebra: Fix up code warnings

CI found a couple of warnings that needed to be cleaned up.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

qlyoung committed:
bgpd: fix UPDATE debugging for prefixes

Recent patch to support EVPN prefix debugging broke debugging regular
prefixes.

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

qlyoung committed:
bgpd: clean up evpn prefix debugging commands

* Fix memory leaks in debug commands
* Use CLI tokens to properly sanitize input
* Remove unnecessary error handling code from debug commands
* Remove unnecessary parameters from conf print functions

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

cfra committed:
isisd: fix refcounting in isis_route.c

This fixes multiple issues and inefficiencies regarding the usage of
route_tables in isis_route.c and removes some memory leaks.

Signed-off-by: Christian Franke <chris@opensourcerouting.org>

rzalamena committed:
bgpd/ospf(6)d/pimd: hide BFD commands with timers

Don't show BFD commands with timers since it might confuse users
("show running-config" won't display timers in client daemons anymore),
but keep accepting this command from previous configurations.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

qlyoung committed:
bgpd: fix odd memory mgmt in clist strings fn

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

qlyoung committed:
tools: improve checkpatch slicing

checkpatch cuts from the diff between the outputs of pre-patch and
post-patch runs of `checkpatch.pl`, but fixed-length greps sometimes
don't cut correctly.

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

lyq140 committed:
ripd: fix packet sending

fix a bug when sending a rip packet.
in authenticate mode but without any string,
no packet should send.

Signed-off-by: lyq140 <34637052+lyq140@users.noreply.github.com>

lyq140 committed:
ripd: fix packet sending in authenticate mode

fix a bug when sending packets.
in authenticate mode but without any string,
no packet should send.

Signed-off-by: lyq140 <34637052+lyq140@users.noreply.github.com>

Commit:
bgpd pimd: return check (Coverity 1472232 1472234)

Signed-off-by: F. Aragon <paco@voltanet.io>

netravnen committed:
doc/user/routemap.rst

qlyoung committed:
Merge pull request #2793 from pacovn/Coverity_1472232_1472234_Unchecked_return_value

bgpd pimd: return check (Coverity 1472232 1472234)

patrasar committed:
PIMD: Fix pim_mroute_del crash while killing pimd.

When we kill pimd, pim_upstream_del() will be get called as part of cleaning process.
Delete mroute by calling pim_mroute_del() which in result initialize the up->channel_oil
to NULL after cleanup. Delete each ifchannel stored in the upstream ifchannel list one
by one. As part of pim_ifchannel_delete(), it will check if it is the last ifchannel,
then delete the corresponding upstream. So pim_upstream_del will be get called more than
once as part of pim_upstream_terminate() and pim_ ifchannel_delete(). Similarly pim_mroute_del()
will be again called and since there is no NULL check before accesing the data, crash is happening.

Fix:
Adding Null check in pim_mroute_del before accessing the data.
and using the FLAG "PIM_UPSTREAM_FLAG_DEL" which will check if the deletion of pim upstream is already
in progress then don’t call pim_upstream_del() from pim_ifchannel_delete().

Signed-off-by: Sarita Patra <saritap@vmware.com>

louberger committed:
Merge pull request #2777 from qlyoung/fix-checkpatch-sh-greps

tools: improve checkpatch slicing

qlyoung committed:
debian: add install-info to build deps

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

qlyoung committed:
debian: remove trailing whitespace from control

Signed-off-by: Quentin Young <qlyoung@cumulusnetworks.com>

rwestphal committed:
Merge pull request #2726 from sworleys/Netlink-Filter-AFI

zebra: Ignore AF_ETHERNET and AF_MPLS in route change as well

ton31337 committed:
bgpd: Display `::` if AFI is IPv6, `0.0.0.0` otherwise

Signed-off-by: Donatas Abraitis donatas.abraitis@gmail.com

eqvinox committed:
Merge pull request #2795 from qlyoung/fix-control-build-deps

Fix control build deps

rzalamena committed:
Merge pull request #2772 from qlyoung/fix-bgp-updates-prefix-debugging

Fix bgp updates prefix debugging

riw777 committed:
Merge pull request #2788 from ton31337/fix/print_ipv6_route_if_afi

bgpd: Display `::` if AFI is IPv6, `0.0.0.0` otherwise

riw777 committed:
Merge pull request #2774 from qlyoung/fix-bgpd-questionable-clist-memory-mgmt

bgpd: fix odd memory mgmt in clist strings fn

Commit:
1. Added json output to "show ip route <prefix>" command
2. Added tag O/P to "show ip route json" command

Signed-off-by: Ameya Dharkar adharkar@vmware.com

Commit:
bgpd: Added local router-id to "show bgp neighbor"

Signed-off-by: Ameya Dharkar adharkar@vmware.com

eqvinox committed:
pimd: fix bogus uninitialized warning

pimd/pim_nht.c: In function ‘pim_ecmp_nexthop_search’:
pimd/pim_nht.c:523:17: error: ‘nbr’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
    nexthop->nbr = nbr;

(on gcc 5.4.0; this is the only warning with that version.)

Signed-off-by: David Lamparter <equinox@diac24.net>

riw777 committed:
Merge pull request #2764 from opensourcerouting/isis-srcdest

IS-IS dst-src routing support

riw777 committed:
Merge pull request #2803 from opensourcerouting/pimd-nbr-uninit-warning

pimd: fix bogus uninitialized warning

riw777 committed:
Merge pull request #2800 from adharkar/frr-bgp_cli

bgpd: Added local router-id to "show bgp neighbor"

eqvinox committed:
staticd: fix tiny whitespace issue

Just add a linebreak.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
build: rework mallinfo test & find malloc_size

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
lib: count total memory allocation per MTYPE

If malloc_usable_size() or malloc_size() are available, we can count
total usage of a particular MTYPE.  (Without the functions, we don't
know how much to subtract on free.)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
doc: add "show memory" user documentation

Signed-off-by: David Lamparter <equinox@diac24.net>

mwinter-osr committed:
Merge pull request #2122 from donaldsharp/zebra_nhs

Create Staticd

srimohans commented on pimd/pim_ifchannel.c:
```diff
@@ -174,8 +174,11 @@ void pim_ifchannel_delete(struct pim_ifchannel *ch)
-	pim_upstream_del(pim_ifp->pim, ch->upstream, __PRETTY_FUNCTION__);
-	ch->upstream = NULL;
+
+        if (!(ch->upstream->flags & PIM_UPSTREAM_FLAG_DEL)) {
```
Starting Review

srimohans replied:
1) Why do we need a separate flag here ? Just checking against ch->upstream is NULL or not would be enough. Am I missing something here ? Given a choice, I would like to avoid the flag as it needs separate maintenance & associated code. 
2) I see an issue here. We are setting ch->upstream = NULL. In case this functions gets called again after ch->upstream set to NULL, then this will crash as we are checking ch->upstream->flag. This causes an NULL pointer exception. This may not have showed up in your testing as there may not be any case where this function gets called after setting ch->upstream = NULL. But this should be avoided.

eqvinox committed:
Merge pull request #2799 from adharkar/frr-zebra_cli

Zebra: Changes to "show ip route" json commands

eqvinox committed:
Merge pull request #2794 from netravnen/feature/docs/set-origin

doc: set origin

eqvinox committed:
Merge pull request #2790 from lyq140/branch1

don't send packets when auth_str is empty (via keychain.)

eqvinox committed:
lib: fix "-t" command line option

was_stdio was getting set for fd == -1 (config file read), thus
prematurely closing the stdio vty.

Signed-off-by: David Lamparter <equinox@diac24.net>

Commit:
configure: fix help text

Correct help text in a couple of places; add configure switch
for new staticd daemon.

Signed-off-by: Mark Stapp <mjs@voltanet.io>

mwinter-osr committed:
bgpd: Fix bgp rpki to match route-map update from PR #2697

Signed-off-by: Martin Winter <mwinter@opensourcerouting.org>

rwestphal committed:
Merge pull request #2807 from mjstapp/configure_help_fixes

configure.ac : fix help text

edipascale committed:
staticd: Fix null pointer access (coverity 1472311)

Aded a check to ensure that if_is_vrf() is not called on a NULL if pointer.
Incidentally this was causing PR #2805 to fail.

Signed-off-by: Emanuele Di Pascale <emanuele@voltanet.io>

eqvinox committed:
Merge pull request #2812 from manuhalo/fix-staticd-coverity

staticd: Fix null pointer access (coverity 1472311)

qlyoung committed:
Merge pull request #2806 from opensourcerouting/vty-term-fd

lib: fix "-t" command line option

dslicenc committed:
tools: correct frr-reload.py handling of bgp vni/exit-vni config

Problem reported that when a peer-group was added in certain
configurations, it would be rejected because of the order of the
commands put in by nclu. Issued turned out to be how frr-reload.py
was handling the sub-sub-context of the vni under the address-family
and subsequently how it handled the following exit-vni.

Ticket: CM-21996
Signed-off-by: Don Slice <dslice@cumulusnetworks.com>

vishaldhingra committed:
peer_delete did not trigger ZEBRA_NEXTHOP_UNREGISTER to ZEBRA

bgp_unlink_nexthop_check() PEER should be NULL and bgp_info list count should be 0

Signed-off-by: vishaldhingra <vdhingra@vmware.com>

eqvinox committed:
tools/checkpatch: fix some bogus macro warnings

checkpatch was throwing an error for "#define FOO , bar"

Signed-off-by: David Lamparter <equinox@diac24.net>

qlyoung committed:
Merge pull request #2809 from opensourcerouting/routemap-rpki-fix

bgpd: Fix bgp rpki to match route-map update from PR #2697

mwinter-osr committed:
doc: Add missing flowspec.rst to distribution tar file

Signed-off-by: Martin Winter <mwinter@opensourcerouting.org>

mwinter-osr committed:
snapcraft: Add new staticd to snap package

Signed-off-by: Martin Winter <mwinter@opensourcerouting.org>

rzalamena committed:
bfdd: fix downtime display

BFD peer status show command was using the wrong variable to display the
downtime counter.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

Commit:
bgpd: return value check (Coverity 1472310)

Signed-off-by: F. Aragon <paco@voltanet.io>

Commit:
staticd: null check (Coverity 1472311)

Signed-off-by: F. Aragon <paco@voltanet.io>

eqvinox committed:
Merge pull request #2824 from vishaldhingra/reg_case

peer_delete did not trigger ZEBRA_NEXTHOP_UNREGISTER to ZEBRA

eqvinox committed:
Merge pull request #2825 from dslicenc/frr-reload-vni

tools: correct frr-reload.py handling of bgp vni/exit-vni config

rzalamena committed:
Merge pull request #2828 from pacovn/Coverity_1472311_null_check

staticd: null check (Coverity 1472311)

qlyoung committed:
Merge pull request #2827 from pacovn/Coverity_1472310_Unchecked_return_value

bgpd: return value check (Coverity 1472310)

Commit:
ospf6d: revert fix for Coverity 1221459

The correction in commit 7edb6aa (PR #2502) was wrong, as it is was not
taking in consideration the unlock counter. Thanks to @eqvinox for noticing
it.

Signed-off-by: F. Aragon <paco@voltanet.io>

qlyoung committed:
Merge pull request #2820 from opensourcerouting/err-prep

remove alloc return value checks + minor babeld cleanup

eqvinox committed:
Merge pull request #2830 from pacovn/Coverity_1221459_revert

ospf6d: revert fix for Coverity 1221459

mwinter-osr committed:
vtysh: Fix missing tailing / in --config_dir option parsing

Signed-off-by: Martin Winter <mwinter@opensourcerouting.org>

rwestphal committed:
zebra: remove unguarded debugging leftovers

These debug messages were committed by accident.

Signed-off-by: Renato Westphal <renato@opensourcerouting.org>

donaldsharp committed:
Merge pull request #2832 from opensourcerouting/rm-debug-leftovers

zebra: remove unguarded debugging leftovers

donaldsharp committed:
Merge pull request #2833 from opensourcerouting/assorted-fixes

Assorted fixes (pre-northbound)

srimohans commented:
I understand the Fix now. Thanks for the meeting and explanation. Approving it.

donaldsharp committed:
Merge pull request #2823 from opensourcerouting/snap-staticd

Updates to snap package to support staticd

donaldsharp committed:
Merge branch 'master' into bfd-final

donaldsharp committed:
Merge pull request #2294 from opensourcerouting/bfd-final

bfdd: add BFD support

ppmathis committed:
bgpd: Always show CIDR mask when displaying routes

Classful networking has been obsolete for ages and there is currently an
inconsistency between `show ip route` and `show bgp`, where the first
one always displays the CIDR mask while the second one hides classful
network masks.

This commit adjusts the behavior of `show bgp` to always show the CIDR
mask for a route, even when it is classful.

Signed-off-by: Pascal Mathis <mail@pascalmathis.com>

qlyoung committed:
Merge pull request #2805 from opensourcerouting/malloc-size

lib: track total memory per MTYPE if possible

rzalamena committed:
bfdd: fix alpine build

Move `recvmsg` parameter option initialization to function, initialize
the parameter struct with zeroes and access field by name.

Signed-off-by: Rafael Zalamena <rzalamena@opensourcerouting.org>

mwinter-osr committed:
snapcraft: Add bfdd to snap package

Signed-off-by: Martin Winter <mwinter@opensourcerouting.org>

louberger committed:
user: document neighbor ... solo config

Signed-off-by: Lou Berger <lberger@labn.net>

donaldsharp committed:
Merge pull request #2843 from LabNConsulting/working/lb/doc-solo

bgpd: document how to avoid reflecting advertised information back to originator

donaldsharp committed:
Merge pull request #2840 from opensourcerouting/snap-bfdd

snapcraft: Add bfdd to snap package

qlyoung committed:
Merge pull request #2837 from ppmathis/cidr-style

bgpd: Always show CIDR mask when displaying routes

donaldsharp committed:
bfdd: Fix `don't assign to value to ourself` compiler warning

We were attempting to get around a unused value by setting
a variable to itself, but it runs afoul of a different
compiler and a message about not setting a variable to
itself.  So let's mark the ttl value as unused.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

rzalamena committed:
Merge pull request #2844 from donaldsharp/unused

bfdd: Fix `don't assign to value to ourself` compiler warning

donaldsharp committed:
lib, ripngd, zebra: Remove pre-solaris 9 special cased code

The CMSG_FIRSTHDR was broken on solaris pre version 9.  Version 9
was released in May of 2002 and EOL'ed in 2014.  Version 8 EOL'ed
in 2012.  Remove special case code for a little used platform
that has not seen the light of day in a very long time.

Signed-off-by: Donald Sharp <sharpd@cumulusnetworks.com>

donaldsharp committed:
Merge pull request #2838 from opensourcerouting/bfd-alpine

bfdd: fix alpine build

eqvinox committed:
Merge pull request #2845 from donaldsharp/cmsg_foolishness

lib, ripngd, zebra: Remove pre-solaris 9 special cased code
